### PR TITLE
[HTML5] Implement Godot <-> JavaScript interface.

### PR DESCRIPTION
**Pull request by Faless**
`4.0` version of: #48691 .

The API is the same, with the exception of `JavaScript.create_callback` which now takes a callable (first line below):

```gdscript
var _my_js_callback = JavaScript.create_callback(myCallback) # This reference must be kept
var console = JavaScript.get_interface("console")

func _init():
	var buf = JavaScript.create_object("ArrayBuffer", 10) # new ArrayBuffer(10)
	print(buf) # prints [JavaScriptObject:OBJECT_ID]
	var uint8arr = JavaScript.create_object("Uint8Array", buf) # new Uint8Array(buf)
	uint8arr[1] = 255
	prints(uint8arr[1], uint8arr.byteLength) # prints 255 10
	console.log(uint8arr) # prints in browser console "Uint8Array(10) [ 0, 255, 0, 0, 0, 0, 0, 0, 0, 0 ]"

	# Equivalent of JavaScript: Array.from(uint8arr).forEach(myCallback)
	JavaScript.get_interface("Array").from(uint8arr).forEach(_my_js_callback)

func myCallback(args):
	# Will be called with the parameters passed to the "forEach" callback
	# [0, 0, [JavaScriptObject:1173]]
	# [255, 1, [JavaScriptObject:1173]]
	# ...
	# [0, 9, [JavaScriptObject:1180]]
	print(args)
```

Not really easy to test in current `master`, but you can build the thread version:
`scons p=javascript target=release_debug threads_enabled=yes`
And check the console. If the startup get stuck, try editing `platform/javascript/js/libs/library_godot_os.js` and have `godot_js_os_hw_concurrency_get` always return 2 (otherwise Godot will spawn to many threads and crash your browser).

### CODE REVIEW

File: doc/classes/JavaScript.xml [hunk 1]
```diff
@@ -11,6 +11,23 @@
...
+			<description>
+			</description>
+		</method>
+		<method name="create_object" qualifiers="vararg">
+			<return type="Variant">
+			</return>
+			<argument index="0" name="object" type="String">
+			</argument>
+			<description>
+				Creates a new javascript object using the [code]new[/code] constructor. The [code]object[/code] must a valid property of the JavaScript [code]window[/code]. See [JavaScriptObject] for usage.
```
Comment (akien-mga) [inline doc/classes/JavaScript.xml:28]:
```suggestion
				Creates a new JavaScript object using the [code]new[/code] constructor. The [code]object[/code] must a valid property of the JavaScript [code]window[/code]. See [JavaScriptObject] for usage.
```

File: doc/classes/JavaScript.xml [hunk 2]
```diff
@@ -11,6 +11,23 @@
 		<link title="Exporting for the Web: Calling JavaScript from script">https://docs.godotengine.org/en/latest/getting_started/workflow/export/exporting_for_web.html#calling-javascript-from-script</link>
 	</tutorials>
 	<methods>
+		<method name="create_callback">
+			<return type="JavaScriptObject">
+			</return>
+			<argument index="0" name="callable" type="Callable">
+			</argument>
+			<description>
```
Comment (akien-mga) [inline doc/classes/JavaScript.xml:19]: Adding a description would be great :)

File: platform/javascript/javascript_singleton.cpp [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+		default:
+			ERR_FAIL_MSG("Unknown lock type to free. Likely a bug.");
+	}
+}
+
+Variant JavaScriptObjectImpl::_js2variant(int p_type, godot_js_wrapper_ex *p_val) {
+	Variant::Type type = (Variant::Type)p_type;
+	switch (type) {
+		case Variant::BOOL:
+			return p_val->i ? Variant(true) : Variant(false);
```
Comment (akien-mga) [inline platform/javascript/javascript_singleton.cpp:175]: `Variant` is probably just as happy if you return `(bool)p_val->i` directly, but this is fine too :)

File: platform/javascript/js/libs/library_godot_javascript_singleton.js [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+		try {
+			obj[key] = GodotJSWrapper.variant2js(p_val_type, p_val_ex);
+			return 0;
+		} catch (e) {
+			GodotRuntime.error(`Error setting variable ${key} on object`, obj);
+			return -1;
+		}
+	},
+
+	godot_js_wrapper_create_object__sig: 'iiiiiiii',
```
Comment (akien-mga) [inline platform/javascript/js/libs/library_godot_javascript_singleton.js:248]: That would be more readable as `viii` ;) (ok I'm pretty sure I already made that joke last year :P)

**Review by akien-mga — APPROVED** [review]

File: doc/classes/JavaScript.xml (hunk 2 above)
Comment (akien-mga) [reply doc/classes/JavaScript.xml:19]: You have one in the `3.x` version:
```
Creates a reference to a script function that can be used as a callback by JavaScript. The reference must be kept until the callback happens, or it won't be called at all. See [JavaScriptObject] for usage.
```

File: platform/javascript/javascript_singleton.cpp (hunk 3 above)
Comment (Faless) [reply platform/javascript/javascript_singleton.cpp:175]: I'll give it a try, as it would skip a an `if`.

Comment (Faless) [reply platform/javascript/javascript_singleton.cpp]: Yep, much better :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**